pkg/storage/etcd/etcd_helper.go

```diff
@@ -290,26 +290,76 @@ func (h *etcdHelper) Set(ctx context.Context, key string, obj, out runtime.Objec
 	return err
 }
 
+func checkPreconditions(preconditions *storage.Preconditions, out runtime.Object) error {
+	if preconditions == nil {
+		return nil
+	}
+	objMeta, err := api.ObjectMetaFor(out)
+	if err != nil {
+		return storage.NewInternalError(fmt.Sprintf("can't enforce preconditions %v on un-introspectable object %v, got error: %v", *preconditions, out, err))
```

Comment:
Optional: code is a bit more readable when you make a storage.NewInternalErrorf that does the sprintf for you, I like to do that personally since it's pretty easy in go.